alien4cloud/alien4cloud.go

```diff
@@ -0,0 +1,2169 @@
+				"application/json",
+			},
+		},
+	)
+
+	if err != nil {
+		return nil, 0, errors.Wrapf(err, "Cannot send a request to get number of logs from application '%s' and environment '%s'", applicationID, environmentID)
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return nil, 0, getError(response.Body)
+	}
+
+	responseBody, err := ioutil.ReadAll(response.Body)
+
+	if err != nil {
+		return nil, 0, errors.Wrapf(err, "Cannot read the body of the log query response '%s' in '%s' environment", applicationID, environmentID)
+	}
+
+	var res struct {
+		Data struct {
+			Data         []Log `json:"data"`
+			From         int   `json:"from"`
+			To           int   `json:"to"`
+			TotalResults int   `json:"totalResults"`
+		} `json:"data"`
+	}
+
+	err = json.Unmarshal(responseBody, &res)
+
+	if err != nil {
+		return nil, 0, errors.Wrap(err, "Unable to unmarshal logs from orchestrator")
+	}
+
+	// Then we send the resquest to get all the logs returned for this deployment.
+
+	logsFilter = logsSearchRequest{
+		From: fromIndex,
+		Size: res.Data.TotalResults,
+		Filters: struct {
+			LogFilter
+			DeploymentID []string `json:"deploymentId,omitempty"`
+		}{LogFilter: filters, DeploymentID: []string{deployments[0].ID}},
+		SortConfiguration: struct {
+			Ascending bool   `json:"ascending"`
+			SortBy    string `json:"sortBy"`
+		}{Ascending: true, SortBy: "timestamp"},
+	}
+
+	body, err = json.Marshal(logsFilter)
+
+	if err != nil {
+		return nil, 0, errors.Wrap(err, "Unable to marshal log filters to get logs for the deployment.")
+	}
+
+	response, err = c.do(
+		"POST",
+		fmt.Sprintf("%s/deployment/logs/search", a4CRestAPIPrefix),
+		body,
+		[]Header{
+			{
+				"Content-Type",
+				"application/json",
+			},
+		},
+	)
+
+	if err != nil {
+		return nil, 0, errors.Wrapf(err, "Cannot send a request to get logs from application '%s' and environment '%s'", applicationID, environmentID)
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return nil, 0, getError(response.Body)
+	}
+
+	responseBody, err = ioutil.ReadAll(response.Body)
+
+	if err != nil {
+		return nil, 0, errors.Wrapf(err, "Cannot read the body of the log query response '%s' in '%s' environment", applicationID, environmentID)
+	}
+
+	err = json.Unmarshal(responseBody, &res)
+
+	if err != nil {
+		return nil, 0, errors.Wrap(err, "Unable to unmarshal logs from orchestrator")
+	}
+
+	return res.Data.Data, len(res.Data.Data), nil
+
+}
+
+////////////////////////////////////////////
+// Methods related to workflow management //
+////////////////////////////////////////////
+
+// RunWorkflow runs a4c workflowName workflow for the given a4cAppID and a4cEnvID
+func (c *a4cClient) RunWorkflow(a4cAppID string, a4cEnvID string, workflowName string) (*WorkflowExecution, error) {
+
+	// The Alien4Cloud endpoint to start a workflow in Alien4Cloud is synchronous and for now, never finishes (Alien4Cloud 2.1.0-SM7).
+	go func() {
```

Comment:
This routine seems to not be stopped in case of timeout below.

Reply:
Maybe using a context may help here.

Reply:
Fixed creating a cancelable context and calling http.NewRequestWithContext()